pkg/provisioner/ironic/ironic.go

```diff
@@ -1525,3 +1538,58 @@ func (p *ironicProvisioner) IsReady() (result bool, err error) {
 	checker := newIronicDependenciesChecker(p.client, p.inspector, p.log)
 	return checker.IsReady()
 }
+
+func (p *ironicProvisioner) IsProvisioning() bool {
+	_, ok := p.getCurrentProvisioningHosts()[p.host.Name]
+	return ok
+}
+
+func (p *ironicProvisioner) CanBeProvisioned() bool {
+	return len(p.getCurrentProvisioningHosts()) < maxProvisioningHosts
+}
+
+func (p *ironicProvisioner) getCurrentProvisioningHosts() map[string]struct{} {
+	//lazy init
+	if p.provisioningHosts != nil {
+		return p.provisioningHosts
+	}
+
+	p.provisioningHosts = make(map[string]struct{})
+	if err := p.loadProvisioningHosts(p.provisioningHosts); err != nil {
+		p.log.Error(err, "Unable to get hosts currently being provisioned")
```

Comment:
Shouldn't this error propagate up through the IsProvisioning() and CanBeProvisioned() APIs?

Reply:
Well, initially I took that approach, just to note that the code in the state machine was slightly more complicated for the standard checks, and still the error handling wasn't very clear (refuse at all the operation?), so I opted for a simplified interface / error management giving priority to state handler (let's process anyway the request).
I also considered that in most of the cases an erroneous situation should be detected by the previous IsReady(), so the chances to get a problem here were fewer (but still non-zero).

Reply:
I agree with Doug here. We should retry otherwise a temporary ironic outage (maybe caused by extreme overload) could cause everything to be allowed to start provisioning.

Reply:
Done